### ISSUE TITLE
md-bottom-nav

### ISSUE THREAD

**Issue by jelbourn** [angular/material2#408]
See https://www.google.com/design/spec/components/bottom-navigation.html

**Comment by mikehaas763** [comment]
Thanks for opening this up @jelbourn. I plan on implementing and have already begun some of the coding. I'm also working on writing up a design doc as you pointed out. There's several more things I need to think about that will hopefully come out in the design document but here's my initial thoughts.

I've been looking at an android/java implementation to learn from that: https://github.com/roughike/BottomBar

So far I've thought about providing a declarative approach such as
```html
<md-bottomnav>
    <md-bottomnav-item>
    </md-bottomnav-item>
</md-bottomnav>
```

and a programatic approach that consumes an array
```html
<md-bottomnav [items]="items">
```

The spec mentions that the component can be used for bottom navigation but it doesn't mention anything about it's behavior automatically positioning itself to the bottom. I'm not sure if it's more in the spirit that using this component should automatically fix it to the bottom of it's containing element, if it shouldn't be concerned with overall layout whatsoever (leaving that up to the consumer of the component), or if it should be a container component that takes up the full height of the view and has a projection area.

**Comment by jelbourn** [comment]
@mikehaas763 

For the API, the first approach is preferable. When you want to include items programmatically, you would use `ngFor` on the item. 

The bottom-nav should behave like the toolbar in WRT to the user being able to decide where to place it.

**Comment by kara** [comment]
Agree with @jelbourn. Also, can we make it `md-bottom-nav` rather than `md-bottomnav`?  Will be easier to read.

**Comment by mikehaas763** [comment]
@jelbourn makes sense, was thinking the same thing about ngFor but wasn't sure what would be more preferable.

@kara I agree. My thinking was that I wasn't sure if it should be consistent with sidenav. However, I think of sidenav as one word but with bottom not so much.

**Comment by mikehaas763** [comment]
WIP design document (not very far yet): https://docs.google.com/document/d/1h3fQaEaWYKv9GISAYitKSjbPvgNBB4_veIo2hOAquOE/edit?usp=sharing

**Comment by mikehaas763** [comment]
Thinking about the API a bit more while working on the design doc, the main piece of the component from a UI perspective is the set of icons. So I have to think about how to use MD icons and could use some feedback.

A few ideas:

- Mimic some of the functionality that's in `MdIcon` in the `MdBottomNavItem` component (perhaps extracting pieces into services). This would mean create the same `@Input()`s, inject `MdIconRegistry`, etc.
- Compose MdIcon within MdBottomNavItem and delegate to it
- Subtype MdIcon with MdBottomNavItem. I haven't test this yet with A2 components and am not sure how well it works or if there are any edge cases related to decorators or other constructs.

Although I'm generally in the composition over inheritance camp as most people are, I think this may be a good use case for it. The nav items are essentially just icons. The only differences I know of being added on in the subtype right now are:

1. optionally displaying a text label underneath
2. communicating with the host element to have the ink effect that's specified in the spec starting at the location of the clicked icon

Thoughts?

**Comment by jelbourn** [comment]
@mikehaas763 The content of the bottom-nav items should be completely up to the user. People will _most likely_ put an `<md-icon>` in there, but the component itself should have no dependency on it.

It should be something like
```html
<md-bottom-nav>
  <a md-nav-item ...> <md-icon>home</md-icon> </a>
  ...
</md-bottom-nav>
```

**Comment by mikehaas763** [comment]
@jelbourn I took this statement from the spec too literally: "Because bottom navigation actions are presented as icons, they should be used for content that can be suitably communicated with icons".

Meaning I thought it was a restriction material2 would want to match. That makes it a lot easier though so I'll just make that a projection area.

**Comment by mikehaas763** [comment]
@jelbourn Do you agree that the component should provide default styling so that if something like the following is used then it will look as it would in the spec?

```html
<a md-nav-item>
  <md-icon>home</md-icon>
  <span>Home</span>
</a>
```

**Comment by jelbourn** [comment]
@mikehaas763 yes, but it should be via an explicit attribute rather than looking for something like `span`. See `md-card` for examples of what I mean.

**Comment by mikehaas763** [comment]
update: I'm doing some prototyping to help flush out some design considerations I'm not sure of yet at mikehaas763/material2 in my bottomnav branch. https://github.com/angular/material2/compare/master...mikehaas763:bottomnav

**Comment by DennisSmolek** [comment]
Forgive my ignorance, but I use a tabs for this exact same functionality... I use CSS to flip the views and you could do the same to hide the indicator. Looking at the design spec's it seems you could use the same code for both components instead of creating something totally different and having to maintain two implementations . 

Here is an example of a simply flipped tabs component: http://plnkr.co/edit/hyacHu?p=preview

Tabs & Bottom Nav: Text or icons, full width, truncation, ripples, 
Tabs: Indicator, scrollable
Bottom Nav: On the bottom
If you don't want the wrapped views you don't have to use the `md-tabs-content` directive

I'm just thinking adding a few attributes or css classes would be better than a whole new component..

**Comment by jelbourn** [comment]
I see the bottom nav as being different from tabs in a few ways:
* It same a slightly different visual presentation
* It is _only_ for navigation (essentially a `<nav>` containing `<a>` elements)
* It is not paginated (which tabs will be eventually)

**Comment by DennisSmolek** [comment]
Right, but with so much being the same I was curious why have two codebases. 
Tabs _can_ be paginated, they don't have to be, especially because bottom navs are recommended to be no more than 4. 

The biggest thing I get is semantically you wouldn't have it be "Tabs" as even though the code is identical the use case is different which I get..

**Comment by michaeljota** [comment]
Hi! Looking in the design doc, you forgot to include that in tablets and desktops this should be displayed aside, to the left. 

btw. I agree this may share some code with tabs, but it would easier to understand if you separate concerns.

**Comment by mikehaas763** [comment]
@michaeljota just for clarity purposes, my understanding of the spec is different.

> Larger displays, like desktop, may achieve a similar effect by using side navigation.

This says to me that a consuming application can _optionally choose_ to use a side nav on bigger displays, but not that the component itself should concern itself with resizing and snapping itself to a side nav.

If it were a valid use case in the future to have a component that did have knowledge to dynamically use a side nav or bottom nav based on screen size, then I'd think that should be a new component that wraps each respectively. Eg. AutoNavigationMenu etc :)

**Comment by michaeljota** [comment]
Yeah, I see your point. I'm spanish speaker, in spanish `may` and `can` are the same, and we almost take them as `should`, so that's on me. 

But you are right, that may it's saying it is optional. This component should not care abour being in a large screen or whatever. :)

**Comment by mikehaas763** [comment]
Does the library already have any utility styling for reseting hyperlinks to essentially no styling? I have to do something similar with this component's default styling.

![image](https://cloud.githubusercontent.com/assets/2096573/19621683/c58408e4-985c-11e6-8556-8c5ce3532e0b.png)

**Comment by pyle** [comment]
Hey all.
Have created a bottom-nav for a project I'm working on. The current implementation I have is as follows and programatic. 
Ideally looking to contribute to the design doc and implenetation since I'm already using material2 for the project. Let me know what you think.

```
<bottom-nav [tabs]="navBarTabs"></bottom-nav>
```
With an interface for the tabs:
```
export interface NavTab {
  title: string;
  icon: string;
  link: string;
}

@Component({
  selector: 'bottom-nav',
  templateUrl: './bottom-nav.ng.html',
  styleUrls: ['./bottom-nav.scss']
})
export class BottomNavComponent {
  @Input() tabs: NavTab[];
}
```

Picture for reference:
![screenshot 2017-06-13 at 11 40 49 am](https://user-images.githubusercontent.com/7279609/27091136-4486f43e-502d-11e7-8cb1-ccc8e1c195a0.png)

**Comment by jelbourn** [comment]
@pyle revisiting this, my current thinking is that the bottom-nav should have the same API as the tab-nav-bar:
```html
<nav md-bottom-nav>
  <a md-bottom-nav-item routerLink="troubleshoot"> ... </a>
  <a md-bottom-nav-item routerLink="status"> ... </a>
  <a md-bottom-nav-item routerLink="faq"> ... </a>
</nav>
```

**Comment by pyle** [comment]
@jelbourn I like that implmentation a lot better, easier to read from the HTML side. I'll start on the design doc sometime next week and open it up for comments

**Comment by robertofd1995** [comment]
Hello all, I have been working on in a component, but this one is using bootstrap 4, if any of you want to do an angular material version, just send a pull request it follows the api that you have been speaking here.


![use-example](https://user-images.githubusercontent.com/6787022/29973333-7eafd5a4-8f2f-11e7-935e-2047b26e063f.gif)

https://github.com/robertofd1995/BottomBar_Component

Thanks also to @pyle , he help me showing me his version

**Comment by pyle** [comment]
Just so I keep everyone updated (and don't lose the doc). I started the [design doc](https://docs.google.com/document/d/1PSYz4YKV82oGGJO9P-dxKhxYYuP-yH5tn9qhKAbWD5A) and have been working on it more as of late.  Time is the factor regarding the progress :)

**Comment by MaciejCaputa** [comment]
Any update on this one? I would be happy to contribute. I have implemented my own custom `mat-bottom-nav` that I use in my application. Let me know what steps would I need to go trough for you to be able to feature it in.

**Comment by pyle** [comment]
Hey @michaeljota I've been flat chat and haven't even worked on this. The first step is finishing the [design doc](https://docs.google.com/document/d/1PSYz4YKV82oGGJO9P-dxKhxYYuP-yH5tn9qhKAbWD5A) so that it's all polished and ready. (I can provide edit access if you'd like to contribute) 
Once that's been reviewed and tweaked the implementation can be done.

Out of curiosity, is your implementation similar to Jeremy's comment above?

**Comment by MaciejCaputa** [comment]
Thanks for answering @pyle.

There are couple of ways to provide the API and I think that there should be some flexibility in how developer can provide data for menu.

1. One option is to provide `@Input() items` (I use it in in one of my projects since for different user roles we have different menus) on`mat-bottomnav` where items is an array of objects with icon, label, and absolute route as properties. Then the menu is rendered internally. Moreover additional `@Input` properties on `mat-bottomnav` such as background color, icon color. According to material guidelines user may or may not wish to display labels for non-active items.
```
<mat-bottomnav tabs="tabs"></mat-bottomnav>
```
2. Other more declarative approach would be to define menu using `mat-bottomnav` and `mat-bottomnav-item` with latter having `icon`, `label`, `route` as input properties.
```
<mat-bottomnav text="hide">
  <mat-bottomnav-item route="/home" icon="label" label="Home">
  </mat-bottomnav-item>
</mat-bottomnav
```
3. Even more declarative approach would be to do something like this:
```
<mat-bottomnav color="primary">
  <mat-bottomnav-item route="/home">
    <mat-icon>home</mat-icon>
    Home
  </mat-bottomnav-item>
</mat-bottomnav
```

As you can see there are things that need to be decided on. I can document it in the docs once you give me the edit access. Then we can decide on an approach and formalise the API.

There will be also more questions to answer such as do we allow `mat-bottomnav` to have other number of items than between 3 and 5. Should it be responsive and be displayed on the side on larger screens as per material guidelines.

**Comment by michaeljota** [comment]
If you read the comments above, the latest example seems to be closes to what is preferred by the Material team, because it would be up to the user what, and how, to put inside the buttons. 

I think this should be close to the current implementation of other navigation components.

**Comment by kenji-1996** [comment]
Curious has to how this is going, I am not sure if I should just design my own, use a proof of concept posted here or wait and use an alternative

**Comment by pyle** [comment]
Apologies to people who are waiting for this. 
I haven't finished the design doc and final spec, but started writing the code this weekend to test out some of the implementation ideas. I hope to have it in master before Q1 ends :smile:

**Comment by mak3** [comment]
When it will be done, it will also apply to Material1?

**Comment by pyle** [comment]
I've written alpha code in my fork that works, but still need to test etc;  And it won't apply to Material1, only Material2 since it's for Angular not AngularJS

**Comment by TobiasMalikowski** [comment]
@pyle can you tell us when this will be done? That kind of navigation is a must have for many people and everyone is currently building a custom solution.

**Comment by baldo12** [comment]
News?

**Comment by Martin-Andersen** [comment]
React Material UI have this and more .. https://material-ui.com/demos/bottom-navigation/

**Comment by pyle** [comment]
Hey all, sorry for the silence but I've been pretty busy as of late (and the times before).
While it has been a while since an update and since i've actually worked on this please note that I don't work on material full time (nor am i developer fulltime) and can't give as many cycles as I'd like without coding 24/7/365 :)

That being said, im going to dig up my previous code and see where I got up to. From memory I had it working no issues but needed to add some tests and polish up some of the styles.

**Comment by GoNode5** [comment]
https://twitter.com/stephenfluin/status/1054499414461472768
Why is @StephenFluin requesting this, while this one still on the list ?

**Comment by StephenFluin** [comment]
@GoNode5 Hey there! We use multiple sources of information to try to prioritize between technical debt, refactoring, new features, new components, bugfixing, API changes, and more. Not everyone reads our GitHub issues, so combining conversations with developers, information from social media channels, and our GitHub issues & PRs gives us a more holistic picture of developer needs and wants.

**Comment by max-SS** [comment]
**App bars: bottom**
Are there plans to give native support to this component?
[Material Guidelines](https://material.io/design/components/app-bars-bottom.html#)
![image](https://user-images.githubusercontent.com/9059596/47575154-c6f1a300-d917-11e8-83f4-a8f9fff418d6.png)

**Comment by Splaktar** [comment]
@max-SS that's offtopic for this thread on bottom navigation.

There is [an issue](https://github.com/angular/material2/issues/13964) for [App bars: top](https://material.io/design/components/app-bars-top.html). But I don't see one for [App bars: bottom](https://material.io/design/components/app-bars-bottom.html) yet. Would you like to open a feature request?

**Comment by max-SS** [comment]
Ok thank you very much for the correction.

**Comment by Spawnrad** [comment]
any news please guys?

**Comment by DenSchiep** [comment]
Are there any updates on this component? You guys should really prioritize this, a lot of people are waiting for this since 2016(?!)...

**Comment by basicBrogrammer** [comment]
This has been open for 3 years?

**Comment by Splaktar** [comment]
Please feel free to click the 👍 in the OP to express your interest in this feature. It is currently a P5, the lowest priority and it is listed in the repo's README that it is "not planned".

If you need this functionality soon, please look at creating your own implementation and possibly contributing it to an open source project like https://github.com/tiaguinho/material-community-components.

**Comment by danieldaeschle** [comment]
+ me waiting for this component

**Comment by breningham** [comment]
So, i've got a bottombar written in Angular, using Angular Material Theming. 

Here's a Stackblits: https://stackblitz.com/edit/mat-bottom-bar
Demo: https://mat-bottom-bar.stackblitz.io/ 

@Splaktar reckon this could be a candidate for the community components?

**Comment by Splaktar** [comment]
@breningham as mentioned in https://github.com/angular/components/issues/408#issuecomment-437511310, [bottom navigation](https://material.io/design/components/bottom-navigation.html) is different from [App bars: bottom](https://material.io/design/components/app-bars-bottom.html). However, I think that would be a good contribution to https://github.com/tiaguinho/material-community-components. I think that the shadow should be on top, rather than bottom, for a bottom app bar.

**Comment by Splaktar** [comment]
Related: MDC's [issue](https://github.com/material-components/material-components-web/issues/59) for tracking the bottom navigation feature.

I've opened https://github.com/angular/components/issues/16367 to track the creation of a `mat-bottom-app-bar`.

**Comment by jlandsmann** [comment]
New year, new component? Any updates here?

**Comment by Splaktar** [comment]
@jlandsmann no updates here. This P5 issue is very low priority at this time. 
It's likely that MDC would need to implement https://github.com/material-components/material-components-web/issues/59 (which hasn't had any movement on it) before one would be added to Angular Material.

**Comment by karocksjoelee** [comment]
Any updates?

**Comment by FANMixco** [comment]
This seems like an option: https://github.com/kevlatus/ngx-bottom-nav#readme

**Comment by f-mon** [comment]
hi are there any plans to add this component?, i've try to read the whole conversation but it's not clear to me, what is the official angular-material team position about this feature request:
1) will be implemented? (if so there are plans)
2) will not be implemented but there is an official suggested workoround (ex: use tabs)
3) other...
Please someone of the angular material team could clarify? 
Thank you for your works!
